<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-задачи" data-toc-modified-id="Описание-задачи-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Описание задачи</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Описание данных</a></span></li></ul></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Обработка-данных" data-toc-modified-id="Обработка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Обработка данных</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обоснование" data-toc-modified-id="Обоснование-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обоснование</a></span><ul class="toc-item"><li><span><a href="#Формула-при-умножении-на-обратимую-матрицу" data-toc-modified-id="Формула-при-умножении-на-обратимую-матрицу-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Формула при умножении на обратимую матрицу</a></span></li><li><span><a href="#Использование-свойств-обратных-и-транспонированных-матриц" data-toc-modified-id="Использование-свойств-обратных-и-транспонированных-матриц-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Использование свойств обратных и транспонированных матриц</a></span></li><li><span><a href="#Использование-свойств-единичных-матриц" data-toc-modified-id="Использование-свойств-единичных-матриц-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>Использование свойств единичных матриц</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.2.4"><span class="toc-item-num">2.2.4&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Проверка-обоснования" data-toc-modified-id="Проверка-обоснования-2.2.5"><span class="toc-item-num">2.2.5&nbsp;&nbsp;</span>Проверка обоснования</a></span></li></ul></li></ul></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Написание-функции" data-toc-modified-id="Написание-функции-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Написание функции</a></span></li><li><span><a href="#Проверка-функции" data-toc-modified-id="Проверка-функции-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Проверка функции</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

### Описание задачи

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

### Описание данных

`/datasets/insurance.csv`

Признаки: пол, возраст и зарплата застрахованного, количество членов его семьи.

Целевой признак: количество страховых выплат клиенту за последние 5 лет.

In [1]:
# импорт библиотек

import pandas as pd

import numpy as np

from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression

from sklearn.metrics import r2_score, mean_squared_error

## Загрузка данных

### Обзор данных

In [2]:
try:
    data = pd.read_csv('datasets/insurance.csv')
except FileNotFoundError as e:
    print(repr(e))
    data = pd.read_csv('/datasets/insurance.csv')

FileNotFoundError(2, 'No such file or directory')


In [3]:
# получаю общую сводку
data.info()

# проверю на корректный вывод
display(data.sample(5))

# проверяю имена колонок на пробелы
display(data.columns)

# проверю на явные дубликаты
print(f'Количество явных дубликатов: {data.duplicated().sum()}')

# описание
data.describe().T

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
1140    1     34.0   26900.0            0                  0
2512    1     19.0   43200.0            1                  0
3693    1     30.0   41500.0            1                  0
2567    0     47.0   32800.0            0                  1
2494    1     42.0   31000.0            0                  1

Index(['Пол', 'Возраст', 'Зарплата', 'Члены семьи', 'Страховые выплаты'], dtype='object')

Количество явных дубликатов: 153


count        mean          std     min      25%      50%  \
Пол                5000.0      0.4990     0.500049     0.0      0.0      0.0   
Возраст            5000.0     30.9528     8.440807    18.0     24.0     30.0   
Зарплата           5000.0  39916.3600  9900.083569  5300.0  33300.0  40200.0   
Члены семьи        5000.0      1.1942     1.091387     0.0      0.0      1.0   
Страховые выплаты  5000.0      0.1480     0.463183     0.0      0.0      0.0   

                       75%      max  
Пол                    1.0      1.0  
Возраст               37.0     65.0  
Зарплата           46600.0  79000.0  
Члены семьи            2.0      6.0  
Страховые выплаты      0.0      5.0

In [4]:
# возможно признаки Возраст и Зарплата содержат только целые числа

data.query('Возраст % 1 != 0')

Empty DataFrame
Columns: [Пол, Возраст, Зарплата, Члены семьи, Страховые выплаты]
Index: []

In [5]:
data.query('Зарплата % 1 != 0')

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
80      1     41.0   32700.0            0                  0
159     1     19.0   64900.0            2                  0
313     0     25.0   32700.0            1                  0
392     1     43.0   32700.0            1                  1
448     1     32.0   32700.0            2                  0
772     1     20.0   65100.0            1                  0
820     0     29.0   32200.0            2                  0
872     0     23.0   32200.0            3                  0
911     0     30.0   32200.0            0                  0
963     1     20.0   64100.0            0                  0
1020    1     36.0   32700.0            2                  0
1078    0     34.0   32200.0            3                  0
1336    1     19.0   32700.0            0                  0
1485    1     41.0   32700.0            0                  0
1486    1     50.0   32200.0            2                  2
1847    1     31.0   32700.0            0                  0
1981    0     21.0   32200.0            1                  0
2191    1     22.0   32200.0            1                  0
2287    1     40.0   32200.0            3                  0
2342    1     25.0   32700.0            2                  0
2385    1     36.0   32700.0            1                  0
2439    1     44.0   32200.0            1                  1
2443    1     25.0   32700.0            0                  0
2805    0     47.0   32700.0            1                  1
2994    1     29.0   32200.0            1                  0
3063    0     24.0   64400.0            0                  0
3139    1     43.0   32200.0            2                  1
3352    1     23.0   32200.0            5                  0
3466    0     22.0   32700.0            4                  0
3499    1     22.0   65100.0            4                  0
3614    1     24.0   32700.0            1                  0
3717    0     34.0   32700.0            0                  0
3788    0     46.0   32700.0            2                  1
3824    1     37.0   32700.0            2                  0
3938    1     23.0   32200.0            1                  0
4182    0     27.0   32200.0            2                  0
4520    0     18.0   32700.0            3                  0
4833    0     24.0   32200.0            1                  0
4935    1     19.0   32700.0            0                  0
4998    1     22.0   32700.0            3                  0

Похоже что есть знаки после одной десятой, ими можно пренебречь.

Предварительные замечания:
- Признаки Возраст и Зарплата содержат только целые числа, можно привести к целочисленному типу;
- Обнаружено 153 явных дубликата. Так как не стоит задача повышения качества обучения то дубликаты можно проигнорировать.

Иных замечаний не обнаружено. Данные готовы к обработке.

### Обработка данных

In [6]:
# приведу признаки к целочисленному типу
data[['Возраст', 'Зарплата']] = data[['Возраст', 'Зарплата']].astype(int)

In [7]:
# подготовлю призанки и целевой признак
features = data.drop(['Страховые выплаты'], axis=1)
target = data['Страховые выплаты']

### Вывод

По итогам обзора и предварительной обработки данных все признаки приведены к целочисленному типу. Выборка разделена на признаки и целевой признак.

## Умножение матриц

### Подготовка

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии?

**Ответ:** При умножении признаков на обратимую матрицу качество линейной регрессии не изменится.

### Обоснование

#### Формула при умножении на обратимую матрицу

При умножении матрицы $X$ на матрицу $P$ получу матрицу $C$:

$$C = XP$$

Формула предсказания для матрицы $C$:
$$a{_1} = Cw{_1}$$

Формула обучения для матрицы $C$:
$$w{_1} = (C^T C)^{-1} C^T y$$

Матрица $B$ это произведение матрицы $X$ на матрицу $P$:
$$w{_1} = ((XP)^T XP)^{-1} (XP)^T y$$

Следовательно формула предсказания:
$$a{_1} = XP((XP)^T XP)^{-1} (XP)^T y$$

#### Использование свойств обратных и транспонированных матриц

Транспонированное произведение матриц равно произведению транспонированных матриц, взятых в обратном порядке:
     $$(AB)^T = B^TA^T$$

Раскрою транспонирование матриц:

$$a{_1} =   XP(P^T X^T XP)^{-1} P^T X^T y$$

Если квадратные матрицы $A$ и $B$ порядка $n$ имеют обратные матрицы, то и их произведение имеет обратную матрицу:
$$(AB)^{-1} = B^{-1}A^{-1}$$

Матрицы ассоциативны, то есть их можно перемножать независимо от того что стоит левее и правее перемножаемых матриц:
$$A(BC) = (AB)C$$

Можно скомбинировать эти правила:
$$(ABC)^{-1} = ((AB)C)^{-1}$$

Следовательно:
$$
(P^T X^T XP)^{-1} = ((P^T X^T X)P)^{-1} = P^{-1} (P^T X^T X)^{-1}
$$

$$
P^{-1} (P^T X^T X)^{-1} = P^{-1} (P^T (X^T X))^{-1} = P^{-1} (X^T X)^{-1} (P^T)^{-1}
$$

$$a{_1} =   X P P^{-1} (X^T X)^{-1} (P^T)^{-1} P^T X^T y$$

#### Использование свойств единичных матриц

При умножении матрицы на обратную ей получается единичная матрица:
$$AA^{-1}=A^{-1}A=E$$

Следовательно можно избавиться от $P P^{-1}$ и  $(P^T)^{-1} P^T$:
$$
a{_1} = X P P^{-1} (X^T X)^{-1} (P^T)^{-1} P^T X^T y = X (X^T X)^{-1} X^T y
$$

#### Вывод

$$a{_1} = X (X^TX)^{-1}X^T y = a$$ 

Доказано что при умножении признаков на обратимую матрицу качество линейной регрессии не изменится.

#### Проверка обоснования

In [8]:
# создам матрицы и векторы согласно описанию
# добавлю вектор соответствующий w0
X = np.concatenate((np.ones((features.shape[0], 1)), features), axis=1)
y = target
w = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)
# получу предсказание и проверю MSE
a = X @ w
display(mean_squared_error(a, y))

0.12334688937098945

In [9]:
C = np.concatenate((np.ones((features.shape[0], 1)), np.dot(features, np.random.normal(size=(4, 4)))), axis=1)
w1 = np.linalg.inv(C.T.dot(C)).dot(C.T).dot(y)
a1 = C @ w1
display(mean_squared_error(a1, y))

0.12334688937099227

In [10]:
# сравню результат с линейной регрессией за счет метода библиотеки sklearn
model = LinearRegression()
model.fit(X,y)
prediction = model.predict(X)
display(mean_squared_error(prediction, y))

0.12334688937098945

Подтверждено что при умножении признаков на обратимую матрицу качество линейной регрессии не изменяется. Минимальные отличия связаны с особенностью хранения дробных чисел.

## Алгоритм преобразования

`Предложите алгоритм преобразования данных для решения задачи. Обоснуйте, почему качество линейной регрессии не поменяется.`

**Алгоритм**

1) Используем `np.random.normal` для получения случайной матрицы $P$ с размерностью $(nxn)$ равной количеству признаков в матрице признков $X$.

2) Используя `np.linalg.inv` проверяем обратима ли матрица $P$. Если да то переходим к шагу 3. Если нет то возвращаемся к шагу 1.

3) Сохраняем полученную через `np.linalg.inv` матрицу $P^{-1}$ для последующей расшифровки признаков.

4) Умножаем матрицу признаков $X$ на обратимую матрицу $P$ и получаем зашифрованную матрицу признаков $C$.

5) При неоходимости получения оригинальных значений признаков умножаем зашифрованную матрицу признаков на сохраненную матрицу $P^{-1}$.

**Обоснование**

При умножении матрицы $А$ с размерностью $(mxn)$ на матрицу $В$ с размерностью $(nxr)$ получаем матрицу с размерностью $(mxr)$.

Следовательно при умножении матрицы $X$ с размерностью $(mxn)$ на матрицу $P$ с размерностью $(nxn)$ получим матрицу $C$ с размерностью $(mxn)$, то есть размерностью повторяющей размерность матрицы $X$.

При умножении на обратную матрицу $P^{-1}$ сохранится та же размерность.

## Проверка алгоритма

###  Написание функции

In [11]:
# подаю на вход функции признаки и случайное число для фиксации результата
def encode_data(features, seed):
    n = features.shape[1] # опредеяю размерность кодирующей матрицы
    np.random.seed(seed) # фиксирую результат
    encoder = np.random.normal(size=(n, n)) # получаю кодирующую матрицу
    
    while np.linalg.det(encoder) == 0: # если кодирующая матрица не обратима то +1 к случайному число до получения обратимой
        seed += 1
        np.random.seed(seed)
        encoder = np.random.normal(size=(n, n))
        
    decoder = np.linalg.inv(encoder) # получаю декодирующую матрицу
    encoded_features = np.dot(features, encoder) # получаю закодированные признаки
    return encoded_features, decoder

Проверю работоспособность функции зашифровав и расшифровав признаки.

In [12]:
# исходные признаки
features.head()

Пол  Возраст  Зарплата  Члены семьи
0    1       41     49600            1
1    0       46     38000            1
2    0       29     21000            0
3    0       21     41700            2
4    1       28     26100            0

In [13]:
# получу зашифрованные занные и декодер
encoded_features, decoder = encode_data(features, 12345)
pd.DataFrame(encoded_features).head()

0             1             2             3
0  38224.186641  61881.000420  49961.234837 -64280.684721
1  29313.558467  47428.845564  38278.822267 -49242.555394
2  16206.481556  26215.538233  21153.670838 -27212.472653
3  32110.072445  52006.047856  42004.453110 -54044.730722
4  20126.326163  32571.440926  26289.724215 -33824.037786

In [14]:
# приведу зашифрованные данные к исходному виду. Топорно, но наглядно.
pd.DataFrame(encoded_features @ decoder, columns=data.columns[0:4]).round().astype(int).head()

Пол  Возраст  Зарплата  Члены семьи
0    1       41     49600            1
1    0       46     38000            1
2    0       29     21000            0
3    0       21     41700            2
4    1       28     26100            0

### Проверка функции

In [15]:
# разделю данные на обучающую и тестовую выборки
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.25, random_state=12345)

In [16]:
model = LinearRegression()
model.fit(features_train, target_train)
prediction = model.predict(features_test)
print('r2 для оригинальных признаков:', r2_score(target_test, prediction))

r2 для оригинальных признаков: 0.4352275684083322


Использую ранее закодированные признаки.

In [17]:
enc_features_train, enc_features_test, enc_target_train, enc_target_test = train_test_split(encoded_features, target, test_size=0.25, random_state=12345)

In [18]:
model.fit(enc_features_train, enc_target_train)
enc_prediction = model.predict(enc_features_test)
print('r2 для зашифрованных признаков:', r2_score(enc_target_test, enc_prediction))

r2 для зашифрованных признаков: 0.4352275684080851


### Вывод

Метрика r2 не отличается между моделью обученной на оригинальных и зашифрованных данных. Функция для шифрования данных клиентов методом матричного умножения работает корректно.

## Общий вывод

Было дано задание разработать метод преобразования данных для защиты персональных данных клиентов страховой компании.  
По итогам работы составлена и проверена на работоспособность функция использующая метод матричного умножения для преобразования данных. Функция может быть дополнена с целью выгрузки декодирующей матрицы в файл для последующего хранения, чего не требовалось по условию задачи.

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
    
В целом, работа очень хорошая. Все задачи в рамках проекта у тебя получились. Ты молодец! Нужно только поправить доказательство.
    
Жду твоих исправлений 😊
</font></b>
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
Отлично! Критических замечаний больше нет. Проект принят! 👏 Поздравляю и желаю удачи!
    
Оставлю несколько дополнительных материалов. Обзоры по применению криптографии в машинном обучении: https://arxiv.org/pdf/1902.04109.pdf, https://www.researchgate.net/publication/342746786_Some_Applications_of_Machine_Learning_in_Cryptography
Подробнее про аналогичный алгоритм: https://www.cs.cmu.edu/~rjhall/JOS_revised_May_31a.pdf
</font></b>
</div>